htlcswitch: Disallow circular routes on same channel

carlaKC opened the pull request:
This PR adds some protection against circular liquidity attacks, as described in #3771. 

Two types of protection against circular routes are identified:
1. Same channel circles: disallow htlcs with the same incoming and outgoing channel ID

Update: after discussion, decided to only implement (1), which simplifies the PR.
~2. Pair channel circles: disallow forwarding of htlcs over a pair of channels when a htlc has already forwarded a htlc with that hash in the opposite direction~

Context for not implementing (2):
1. Effect on base MPP: base MPP uses duplicate htlcs, which are indistinguishable from circular routes (in almost all cases, except for payments which reverse over the same channel pair)
2. Pair channel effectiveness: pair channel protection is probably quite easy to get around, so not worth implementing

carlaKC commented:
Quite simple change if we only disallow circles in the same channel. 
Started working on an itest for this as well, but unsure if it's worth adding (in terms of adding more for travis to do).

joostjager commented on config.go:
```diff
@@ -341,6 +341,8 @@ type config struct {
 
 	LegacyProtocol *lncfg.LegacyProtocol `group:"legacyprotocol" namespace:"legacyprotocol"`
+
+	AllowCircularRoute bool `long:"allow-circular" description:"If true, our node will allow htlc forwards that arrive and depart on the same channel."`
```
Match var name and flag name? Is there a use case for allowing this btw?

carlaKC replied:
I think that keysend will replace the need for this eventually, but I'm worried about this change breaking no-invoice apps which tip by dropping off fees (could be the case that they drop off over the same channel for single channel nodes) in the meantime.

joostjager commented on htlcswitch/link.go:
```diff
@@ -2144,8 +2144,8 @@ func (l *channelLink) UpdateForwardingPolicy(newPolicy ForwardingPolicy) {
-	incomingTimeout, outgoingTimeout uint32,
-	heightNow uint32) *LinkError {
+	incomingTimeout, outgoingTimeout, heightNow uint32,
+	incomingLink lnwire.ShortChannelID, allowCircular bool) *LinkError {
```
Maybe it is better to get this static config bool from the link configuration, rather than passing it every time.

joostjager replied:
But I actually think that this check is something that should be done in the switch, when the forwarding decision is made.

carlaKC replied:
I thought about putting it in the switch (in `handlepacketforward`) but then it's less easy to unit test. It's quite a simple change though so can move it in there as well, which saves having to get that bool down to the link?

carlaKC commented:
Moved check into the switch and added an itest. 

The downside of moving this check into the switch is that it exposes the switch to link level information in the form of channelupdate/temporary failure. Instead of adding a `ChannelUpdate` function to `ChannelLink` interface, I could add `TemporaryFailure` which returns the correct error?

joostjager commented on htlcswitch/switch.go:
```diff
@@ -1075,6 +1079,23 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+		// whether the htlc will produce a circular forward through the
+		// same channel, and fail it if the switch does not allow
+		// forwards of this nature.
+		if linkErr := s.CheckCircularForward(
```
A separate statement to call `CheckCircularForward` rather than embedding it in the if condition is probably more readable.

joostjager commented on htlcswitch/switch.go:
```diff
@@ -1170,6 +1191,47 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+	}
+
+	return NewDetailedLinkError(
+		lnwire.NewTemporaryChannelFailure(update),
```
I don't think we need to return an update. There isn't a problem with the policy.

carlaKC replied:
Nil update is fine because the spec has a length field, right? It doesn't specifically say whether we have to include an update. 

Also double checking, `TemporaryChannelUpdate` is the correct error to send?
Seems like the most fitting one in the bolts: "The channel from the processing node was unable to handle this HTLC, but may be able to handle it, or others, later."

joostjager replied:
Yes, I think that is ok. It doesn't matter much which error we return, because the sender must know what they are doing.

joostjager commented on lntest/itest/lnd_test.go:
```diff
@@ -5060,6 +5063,104 @@ func testMultiHopSendToRoute(net *lntest.NetworkHarness, t *harnessTest) {
+	// Finally, we want to test allowing and disallowing of circular
+	// payments. Bob's node disallows circular payments and carol's node
+	// allows them. We will send payments over circular routes through
+	// each of these nodes.
```
Wouldn't it be better to test this in a unit test like `TestSkipIneligibleLinksMultiHopForward`? To me it looks like there is nothing that an itest can exercise that can't be exercised in a lower level test.

joostjager commented on htlcswitch/switch.go:
```diff
@@ -1075,6 +1079,23 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+		// whether the htlc will produce a circular forward through the
+		// same channel, and fail it if the switch does not allow
+		// forwards of this nature.
+		if linkErr := s.CheckCircularForward(
```
I am wondering how this check works together with the previous non-strict forwarding decision. Could it be that the same in and out chan was requested, but that the out channel for some reason (concurrent payment perhaps?) doesn't have enough balance and another channel to the same node is selected?

carlaKC replied:
I think that it's fair to treat these payments as 'suspicious' because they are taking an inefficient route. While there may be some custom use cases out there, node operators who are expecting to use them should configure their node to allow circular routes. 

If somebody tries to route a payment over the same channel (and we do not allow that), we could just catch that early and not apply non-strict forwarding? Which would just mean moving the check up in `handlePacketForward` and failing early.

joostjager replied:
Yes moving it up sounds good. Or maybe basing the check on a pubkey comparison, maybe that is even better.

carlaKC replied:
Disallowing based on pubkey would mean that people can't do 2 hop circular rebalances, which I think would cause some issues for existing nodes if their lnd peers suddenly stopped letting them rebalance.

joostjager replied:
Two hop circular rebalances are problematic anyway, because of non-strict forwarding. You may just as well get back your money through the same channel.

Comparing on the exact chan ids is also a risk for peers that have multiple channels between them. The ping pong route can still be executed. I think it is more important to prevent that than to allow the two hop rebalance which isn't reliable anyway.

carlaKC replied:
It is unreliable but I think a lot of people rely on it at the moment. Would say that we'd probably disable by default if we disallow to the same peer, so that we don't break a lot of people's setups. 

I think that single channel is still a valuable defense to have. Without it, an attacker can lock up funds in every channel on your node by zig zagging back and fourth through each channel. The takes `capacity/(max htlc size*19)` payments to lock up your balance (assuming one hop where they lock up their own balance to route the first hop to your node. I think that to one peer would need wider discussion, @alexbosworth has an operational requirement for that ability, for example.

joostjager replied:
Yes, single channel is better than nothing. I have no data myself on how many people rely on two hop rebalancing. Detecting peer-level ping-pong can be implemented later if there are doubts.

One way to lock up the channel is to use up all its capacity, but another way is to use up all commitment tx slots. There is a max number of htlcs that the tx can hold. That way it costs only negligible money to do the attack.

bhandras commented on htlcswitch/switch.go:
```diff
@@ -1170,6 +1194,40 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+	// Check that the incoming link is different to the outgoing link.
+	// If they are equal, the htlc is part of a circular route which may
+	// be used to lock up our liquidity.
+	isCircularRoute := incoming.ToUint64() == outgoing.ToUint64()
```
I think the `ToUint64()` is not required here as these structs are comparable and comparison should actually be faster too.

bhandras commented on htlcswitch/switch.go:
```diff
@@ -1170,6 +1194,40 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+
+	// If the route is not circular we do not need to perform any further
+	// checks.
+	if !isCircularRoute {
```
init: Maybe further simplify this to `if incoming != outgoing`.

bhandras commented on htlcswitch/switch.go:
```diff
@@ -1170,6 +1194,40 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+	// If the switch is configured to allow circular routes, log that we
+	// are allowing the route then return nil.
+	if allowCircular {
+		log.Info("allowing circular route over link: %v "+
```
Do we want to info log each time this happens? Maybe a debug log is enough.

carlaKC replied:
Wasn't expecting it to happen too often, since it's the case where somebody has changed their config and there's a strange route going through their node, but downgraded to debug. Would imagine that operators who are using these special case routes are running debug logs anyway.

bhandras commented on htlcswitch/switch.go:
```diff
@@ -1170,6 +1194,40 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+	// are allowing the route then return nil.
+	if allowCircular {
+		log.Info("allowing circular route over link: %v "+
+			"(payment hash: %x)", paymentHash, incoming)
```
`paymentHash` and `incoming` are switched in the format string.

bhandras commented on htlcswitch/switch_test.go:
```diff
@@ -1294,6 +1296,178 @@ type multiHopFwdTest struct {
+			t.Parallel()
+
+			alicePeer, err := newMockServer(
+				t, "alice", testStartingHeight, nil, 6,
```
I know this is a pattern in this file, but it really bothers me that 6 is not a constant :)

carlaKC replied:
nit and thy shall receive! (a separate commit declaring a constant for these tests 😄 )

Thanks for pointing it out, it's always nice to do a little clean up while you're around. I think it would be cool to switch over to functional options for some or these test structs with a sensible default so that exactly what you're spinning up is more clear/controlled. But that's a separate discussion/PR.

bhandras commented on htlcswitch/switch_test.go:
```diff
@@ -1294,6 +1296,178 @@ type multiHopFwdTest struct {
+
+			// Create a new packet that loops through alice's link
+			// in a circle.
+			preimage := [sha256.Size]byte{1}
```
As `preimage` and `rhash` are constants for all sub tests, you may consider moving them out from the loop.

bhandras commented on htlcswitch/switch_test.go:
```diff
@@ -1294,6 +1296,178 @@ type multiHopFwdTest struct {
+			// If the error is non-nil, check that the error
+			// contains the error string for a circular forward.
+			if err != nil {
+				if !strings.Contains(err.Error(),
```
nit: Could be just one if.

bhandras commented on htlcswitch/switch_test.go:
```diff
@@ -1294,6 +1296,178 @@ type multiHopFwdTest struct {
+			}
+
+			// Ensure that no circuits were opened.
+			if s.circuits.NumOpen() != 0 {
```
nitnitnit :nerd_face: `if s.curcuits.NumOpen() > 0`

joostjager commented on htlcswitch/link_test.go:
```diff
@@ -39,6 +39,7 @@ import (
 
 const (
 	testStartingHeight = 100
+	testDefaultDelta   = 6
```
For commits that only change test code, we often use `htlcswitch/test` as the commit message prefix

joostjager commented on htlcswitch/switch.go:
```diff
@@ -66,6 +66,12 @@ var (
+	// router when htlcs fail due to circular routes. It is declared here
+	// to allow for exact error matching in tests.
+	circularRouteErrTemplate = "incoming HTLC(%x) disallowed due to " +
+		"circular forward over: %v "
```
Downside of formatting is that it is no longer possible to check for this error via `err == circularRouteErr`

carlaKC replied:
I guess this is a more general question of how specific we want the errors we surface to the user to be. If we want to include metadata then I think this is an improvement on unit tests where we just copy the error string that is returned inline. If we're happy with more generic user errors, then the failure detail's error string can just be returned.

joostjager commented on htlcswitch/switch.go:
```diff
@@ -986,6 +996,25 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+			addErr := fmt.Errorf(circularRouteErrTemplate,
+				htlc.PaymentHash[:], packet.outgoingChanID)
+
+			return s.failAddPacket(
```
I am wondering, what is the purpose of returning addErr here (via `failAddPacket`) and not just `linkErr`? `FailureDetail` already describes what is the problem.

carlaKC replied:
The `addrErr` provides the specific channel ID that the circle was attempted on. I've actually just removed this in #3844 though, I think we can get by without it. If we find ourselves wanting specific link errors, we can just make new `FailureDetail` types that have additional metadata.

carlaKC replied:
Going to kill `addrErr` here and just pass the `linkErr` in twice, since the change to remove `addErr` from `failAddPacket` is already in 3844.

cfromknecht commented on htlcswitch/switch.go:
```diff
@@ -1170,6 +1190,37 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+	// switch is configured to allow circular routes, log that we are
+	// allowing the route then return nil.
+	if allowCircular {
+		log.Debugf("allowing circular route over link: %v "+
```
nit: capitalize log message
